openid_connect.js

```diff
@@ -7,7 +7,38 @@ var newSession = false; // Used by oidcAuth() and validateIdToken()
 
 export default {auth, codeExchange, validateIdToken, logout};
 
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function isVarDefined(variable) {
+    const v = await sleep(10);
+    if (variable != null) return true;
+    else return false;
+}
+
+// If the ID token has not been synced yet, poll the variable every 10ms until
+// get a value or after a timeout.
+async function waitUntilVarIsDefined(r, timeout) {
+    let timeWas = new Date();
+    let result = false;
+    do {
+        result = await isVarDefined(r.variables.session_jwt);
+    } while ((! result) && (new Date() - timeWas < timeout));
+    return result;
+}
+
 function auth(r) {
+    // If a cookie was sent but the ID token is not in the key-value database, wait for the token to be in sync.
+    if (r.variables.cookie_auth_token != null && r.variables.session_jwt == null && r.variables.zone_sync_timeout > 0) {
+        waitUntilVarIsDefined(r, r.variables.zone_sync_timeout)
+        .then(result => {
+            if (result) {
+                r.internalRedirect(`${r.variables.uri}${r.variables.is_args ? `${r.variables.is_args}${r.variables.args}` : ``}`);
```

Comment:
Below it's done like this:
```suggestion
                delete r.headersOut["WWW-Authenticate"];
                r.internalRedirect(r.variables.request_uri);
```

Reply:
Perhaps, this should be made a function while fixing "URL encoding", e.g. `retryOriginalRequest`.
Also, redirect implementation could be more eye-friendly, e.g. `r.internalRedirect(r.variables.uri + r.variables.is_args + r.variables.args);`.

Reply:
Good point. We can't use the proposed option because "r.variables.args" can return "undefined", but this works fine:
`r.internalRedirect(r.variables.uri + r.variables.is_args + (r.variables.args || ''));`

Reply:
Still makes sense to deal with this in a separate "URL encoding" commit (and deny '?' in uri there as well).